go/vt/vtgate/engine/mirror.go

```diff
@@ -0,0 +1,160 @@
+/*
+Copyright 2024 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"context"
+	"math/rand/v2"
+	"time"
+
+	"vitess.io/vitess/go/sqltypes"
+	querypb "vitess.io/vitess/go/vt/proto/query"
+)
+
+type (
+	// percentBasedMirror represents the instructions to execute an
+	// authoritative primitive and, based on whether a die-roll exceeds a
+	// percentage, to also execute a target Primitive.
+	percentBasedMirror struct {
+		percent   float32
+		primitive Primitive
+		target    Primitive
+	}
+)
+
+const (
+	// maxMirrorTargetLag limits how long a mirror target may continue
+	// executing after the main primitive has finished.
+	maxMirrorTargetLag = 100 * time.Millisecond
+)
+
+var _ Primitive = (*percentBasedMirror)(nil)
+
+// NewPercentBasedMirror creates a Mirror.
+func NewPercentBasedMirror(percentage float32, primitive Primitive, target Primitive) *percentBasedMirror {
+	return &percentBasedMirror{percentage, primitive, target}
+}
+
+func (m *percentBasedMirror) RouteType() string {
+	return "Mirror"
+}
+
+func (m *percentBasedMirror) GetKeyspaceName() string {
+	return m.primitive.GetKeyspaceName()
+}
+
+func (m *percentBasedMirror) GetTableName() string {
+	return m.primitive.GetTableName()
+}
+
+func (m *percentBasedMirror) GetFields(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable) (*sqltypes.Result, error) {
+	return m.primitive.GetFields(ctx, vcursor, bindVars)
+}
+
+func (m *percentBasedMirror) NeedsTransaction() bool {
+	return m.primitive.NeedsTransaction()
+}
+
+func (m *percentBasedMirror) TryExecute(ctx context.Context, vcursor VCursor, bindVars map[string]*querypb.BindVariable, wantfields bool) (*sqltypes.Result, error) {
+	var mirrorCh chan any
+	var mirrorCtx context.Context
+	var mirrorCtxCancel func()
+
+	doMirror := m.percentAtLeastDieRoll()
+
+	if doMirror {
+		mirrorCh = make(chan any)
+
+		mirrorCtx, mirrorCtxCancel = context.WithTimeout(ctx, maxMirrorTargetLag)
+		defer mirrorCtxCancel()
+
+		go func(target Primitive, vcursor VCursor) {
+			defer close(mirrorCh)
+			_, _ = target.TryExecute(mirrorCtx, vcursor, bindVars, wantfields)
+		}(m.target, vcursor.CloneForMirroring(mirrorCtx))
```

Comment:
should we not log any error on executing this mirror primitive?

Reply:
If the execution returns an error

Reply:
can have 
```
var mirrorCh chan error

send error on the channel

_, err := target.TryExecute(mirrorCtx, vcursor, bindVars, wantfields)
			mirrorCh <- err
```

Reply:
@harshit-gangal what i would ultimately like to do is somehow put errors into logstats. not exactly sure how that will work but it could be as simple as adding a new field like `MirrorError` and setting it here. i would like to tackle that in a separate PR if that's alright with you.